src/neo/SmartContract/Native/Designate/DesignateContract.cs

```diff
@@ -37,29 +38,64 @@ public ECPoint[] GetDesignatedByRole(StoreView snapshot, Role role)
         {
             if (!Enum.IsDefined(typeof(Role), role))
                 throw new ArgumentOutOfRangeException(nameof(role));
-            return snapshot.Storages[CreateStorageKey((byte)role)].GetInteroperable<NodeList>().ToArray();
+            byte[] index = snapshot.Storages[CreateStorageKey((byte)role)].Value;
+            KeyBuilder index_key = CreateStorageKey((byte)role).Add(index);
+            return snapshot.Storages[index_key].GetInteroperable<NodeList>().ToArray();
+        }
+
+        [ContractMethod(0_01000000, CallFlags.AllowStates)]
+        public ECPoint[] GetDesignatedByRoleAndIndex(StoreView snapshot, Role role, uint index)
+        {
+            if (!Enum.IsDefined(typeof(Role), role))
+                throw new ArgumentOutOfRangeException(nameof(role));
+            if (snapshot.Height + 2 < index)
+                throw new ArgumentOutOfRangeException(nameof(index));
+            List<uint> keys = snapshot.Storages.Find(CreateStorageKey((byte)role).ToArray())
+                .Select(p => p.Key.Key.Skip(1).ToArray())
+                .Where(p => p.Length == sizeof(uint))
+                .Select(p => BitConverter.ToUInt32(p))
+                .ToList();
+            keys.Sort();
+
+            if (keys.Count == 0) return System.Array.Empty<ECPoint>();
+
+            uint height = 0;
+            foreach (uint h in keys)
+            {
+                if (index <= h) break;
+                height = h;
+            }
+
+            if (0 < height)
+                return snapshot.Storages[CreateStorageKey((byte)role).Add(height)].GetInteroperable<NodeList>().ToArray();
+
+            return System.Array.Empty<ECPoint>();
         }
 
         [ContractMethod(0, CallFlags.AllowModifyStates)]
-        private void DesignateAsRole(ApplicationEngine engine, Role role, ECPoint[] nodes)
+        private bool DesignateAsRole(ApplicationEngine engine, Role role, ECPoint[] nodes)
         {
             if (nodes.Length == 0 || nodes.Length > 32)
                 throw new ArgumentException();
             if (!Enum.IsDefined(typeof(Role), role))
                 throw new ArgumentOutOfRangeException(nameof(role));
             if (!CheckCommittee(engine)) throw new InvalidOperationException();
-            NodeList list = engine.Snapshot.Storages.GetAndChange(CreateStorageKey((byte)role)).GetInteroperable<NodeList>();
+            uint index = engine.Snapshot.Height + 1;
+            NodeList list = engine.Snapshot.Storages.GetAndChange(CreateStorageKey((byte)role).Add(index), () => new StorageItem(new NodeList())).GetInteroperable<NodeList>();
             list.Clear();
             list.AddRange(nodes);
             list.Sort();
+            StorageItem current = engine.Snapshot.Storages.GetAndChange(CreateStorageKey((byte)role));
+            current.Value = BitConverter.GetBytes(index);
+            return true;
```

Comment:
If we never return `false` we can avoid return `bool`

Reply:
Done.